exp/runtime/hooks/api/v1alpha1/aggregate.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+)
+
+type AggregatableResponse interface {
+	runtime.Object
+
+	Aggregate(items []*ExtensionResponse) (HookResponseSummary, error)
+}
+
+type ExtensionResponse struct {
+	ExtensionName string
+	Response      runtime.Object
+	FailurePolicy *runtimev1.FailurePolicy
+	Error         error
+}
+
+type ExtensionResponseList []*ExtensionResponse
+
+type ResponseSummaryStatus string
+
+const (
+	ResponseSummaryStatusSuccess        ResponseSummaryStatus = "Success"
+	ResponseSummaryStatusFailureIgnored ResponseSummaryStatus = "FailureIgnored"
+	ResponseSummaryStatusFailure        ResponseSummaryStatus = "Failure"
+)
+
+type ExtensionResponseSummary struct {
+	ExtensionName string
+	Status        ResponseSummaryStatus
+	Error         error
+}
+
+type HookResponseSummary []ExtensionResponseSummary
+
+func aggregateExtensionResponses(response runtime.Object, responses []*ExtensionResponse) (HookResponseSummary, error) {
+	msg := &strings.Builder{}
+	status := ResponseStatusSuccess
+	summaries := HookResponseSummary{}
+
+	for _, item := range responses {
+		ignore := *item.FailurePolicy == runtimev1.FailurePolicyIgnore
+		summaryStatus := ResponseSummaryStatusSuccess
+
+		if item.Error != nil {
+			if !ignore {
+				summaryStatus = ResponseSummaryStatusFailure
+				status = ResponseStatusFailure
+				m, err := GetMessage(item)
+				if err != nil {
+					return nil, err
+				}
+				msg.WriteString(m)
+			} else {
+				summaryStatus = ResponseSummaryStatusFailureIgnored
+			}
+		}
+
+		summaries = append(summaries, ExtensionResponseSummary{
+			ExtensionName: item.ExtensionName,
+			Error:         item.Error,
+			Status:        summaryStatus,
+		})
+	}
+
+	if err := SetMessage(response, msg.String()); err != nil {
+		return nil, err
+	}
+	if err := SetStatus(response, status); err != nil {
+		return nil, err
+	}
+	return summaries, nil
+}
+
+func aggregateRetryAfterSeconds(response runtime.Object, responses []*ExtensionResponse) error {
+	// TODO: Ignore responses that are failures.
```

Comment:
Can we assume that as soon as we have an actual failure (after failurePolicy) we wouldn't use retryAfterSeconds anyway?

This means either:
a) we would never call this func when we found an error before. Therefore we can simply aggregate retryAfterSeconds here and don't care about failures
b) at the beginning of this func we check if the response is a failure and then we just return nil without doing anything

Both options imply a certain ordering in which aggregateExtensionResponses and aggregateRetryAfterSeconds are called. I wonder if we should rename aggregateRetryAfterSeconds a bit and then call aggregateExtensionResponses internally. This is probably the only way to avoid duplicating logic or awkward dependencies.